daisi/examples/tcp/tcp_example.cpp

```diff
@@ -0,0 +1,157 @@
+public:
+  static TypeId GetTypeId() {
+    static TypeId tid = TypeId("Server").SetParent<Application>().AddConstructor<Server>();
+    return tid;
+  }
+
+  Server() = default;
+
+  ~Server() override = default;
+
+  void StartApplication() override {
+    auto new_msg = [this](network_tcp::TcpSocketHandle handle, const std::string &msg) {
+      std::cout << "[SERVER] \"" << msg << "\" from handle " << handle << "\n";
+      tcp_->send(handle, "PONG");
+    };
+
+    auto connected = [](network_tcp::TcpSocketHandle handle, const std::string &ip, uint16_t port) {
+      std::cout << "[SERVER] New connection from " << ip << ":" << port << " WITH HANDLE " << handle
+                << "\n";
+    };
+
+    auto disconnected = [](network_tcp::TcpSocketHandle /*handle*/) {};
+
+    tcp_ = std::make_unique<network_tcp::Server>(
+        network_tcp::ServerCallbacks{new_msg, connected, disconnected});
+    std::cout << "[SERVER] Started on " << tcp_->getIP() << ":" << tcp_->getPort() << "\n";
+  }
+
+  void StopApplication() override { tcp_.reset(); }
+
+private:
+  std::unique_ptr<network_tcp::Server> tcp_;
+};
+
+/// TCP client
+class Client : public Application {
+public:
+  static TypeId GetTypeId() {
+    static TypeId tid = TypeId("Client").SetParent<Application>();
+    return tid;
+  }
+
+  explicit Client(network_tcp::Endpoint endpoint) : endpoint_(std::move(endpoint)) {}
+
+  ~Client() override = default;
+
+  void StartApplication() override {
+    auto new_msg = [](const std::string & /*msg*/) {};
+    auto connected = []() {};
+    auto disconnected = []() {};
+
+    tcp_ = std::make_unique<network_tcp::Client>(
+        network_tcp::ClientCallbacks{new_msg, connected, disconnected}, endpoint_);
+  }
+
+  void StopApplication() override { tcp_.reset(); }
+
+  void sendData(const std::string &msg) { tcp_->send(msg); }
+
+private:
+  network_tcp::Endpoint endpoint_;
+
+  std::unique_ptr<network_tcp::Client> tcp_;
+};
+
+#define SCHEDULE(number, function, type, seconds)                             \
+  ns3::Simulator::ScheduleWithContext(                                        \
+      node_container.Get(number)->GetId(), Seconds(seconds), &type::function, \
+      node_container.Get(number)->GetApplication(0)->GetObject<type>());
+
+#define SCHEDULE_WITH_ARG(number, function, type, seconds, arg)               \
+  ns3::Simulator::ScheduleWithContext(                                        \
+      node_container.Get(number)->GetId(), Seconds(seconds), &type::function, \
+      node_container.Get(number)->GetApplication(0)->GetObject<type>(), arg);
+
+int main(int /*argc*/, char * /*argv*/[]) {
+  NodeContainer node_container;
+  node_container.Create(2);
+
+  CsmaHelper csma;
+  csma.SetChannelAttribute("DataRate", StringValue("10Gbps"));
+  csma.SetChannelAttribute("Delay", StringValue("1ms"));
+
+  NetDeviceContainer net_devices;
+  net_devices = csma.Install(node_container);
+  csma.EnablePcapAll("output");
```

Comment:
Since when are we doing PCAP tracing?

Reply:
This is just for the example, so that the output can be inspected.